client/scripts/mifarePlus.lua

```diff
@@ -0,0 +1,310 @@
+		local count,cmd,arg1,arg2,arg3,data = bin.unpack('LLLLH512',result)
+		returned_bytes = string.sub(data, 1, arg1 * 2)
+		print(("<recvd>: %s"):format(returned_bytes)) -- need to multiply by 2 because the hex digits are actually two bytes when they are strings
+		return returned_bytes
+	else
+		err = "Error sending the card raw data."
+		oops(err)
+	end
+end
+
+function writePerso()
+	-- Used to write any data, including the keys (Key A and Key B), for all the sectors.
+	-- writePerso() command parameters:
+	--		1 byte  - 0xA8 - Command Code
+	--		2 bytes - Address of the first block or key to be written to (40 blocks are numbered from 0x0000 to 0x00FF)
+	--		X bytes - The data bytes to be written, starting from the first block. Amount of data sent can be from 16 to 240 bytes in 16 byte increments. This allows
+	--				up to 15 blocks to be written at once.
+	-- response from PICC:
+	--		0x90 - OK
+	--		0x09 - targeted block is invalid for writes, i.e. block 0, which contains manufacturer data
+	--		0x0B - command invalid
+	--		0x0C - unexpected command length
+	
+
+	-- First, set all the data in the card to zeros. The keys, stored in the sector trailer block, are also set to zeros.
+	-- The only block which cannot be explicitly set is block 0x0000, the manufacturer block.
+	print("Setting values of normal blocks")
+	cardsize = 4 --need to set to 4 for 4k or 2 for 2k
+	if(cardsize == 4) then
+		numblocks = 255
+	elseif(cardsize == 2) then
+		numblocks = 127
+	else
+		oops("Invalid card size")
+	end
+
+	for i=1,numblocks,1 do --skip block 0
+		--convert the number to hex with leading zeros, then use it as the block number in writeBlock()
+		blocknum = string.format("%04x", i)
+		writeBlock(blocknum, SIXTEEN_BYTES_ZEROS)
+	end
+	print("Finished setting values of normal blocks")
+
+	print("Setting AES Sector keys")
+	-- Next, write to the AES sector keys
+	for i=0,39 do --for each sector number
+		local keyA_block = "40" .. string.format("%02x", i * 2)
+		local keyB_block = "40" .. string.format("%02x", (i * 2) + 1)
+		--Can also calculate the keys fancily to make them unique, if desired
+		keyA = SIXTEEN_BYTES_ZEROS
+		keyB = SIXTEEN_BYTES_ZEROS
+		writeBlock(keyA_block, keyA)
+		writeBlock(keyB_block, keyB)
+	end
+	print("Finished setting AES Sector keys")
+
+	print("Setting misc keys which haven't been set yet.")
+	--CardMasterKey
+	blocknum = "9000"
+	writeBlock(blocknum, SIXTEEN_BYTES_ZEROS)
+	--CardConfigurationKey
+	blocknum = "9001"
+	writeBlock(blocknum, SIXTEEN_BYTES_ZEROS)
+	--L3SwitchKey
+	blocknum = "9003"
+	writeBlock(blocknum, SIXTEEN_BYTES_ZEROS)
+	--SL1CardAuthKey
+	blocknum = "9004"
+	writeBlock(blocknum, SIXTEEN_BYTES_ZEROS)
+	--L3SectorSwitchKey
+	blocknum = "9006"
+	writeBlock(blocknum, SIXTEEN_BYTES_ZEROS)
+	--L1L3MixSectorSwitchKey
+	blocknum = "9007"
+	writeBlock(blocknum, SIXTEEN_BYTES_ZEROS)
+	--VC Keys
```

Comment:
Missing: 9002 L2SwitchKey

Reply:
Not present in my datasheet (Mifare Plus EV1). Someone can always add it if they need to.

Reply:
Contradicts with my information (non EV1). The none-EV1 came as Mifare Plus X and Mifare Plus S. Are there different Mifare Plus EV1 as well? We should avoid to make the script work for only one out of many Mifare Plus types.

Reply:
Mifare Plus S and Mifare Plus X (and Mifare Plus SE I think) were the earlier versions of Mifare Plus; EV1 is the newest version (with a few different models). It looks like one of the big things they did in EV1 was get rid of security level 2, which is why we have this discrepancy on keys and a couple other minor things. Do you think I should just add something in the header and say the exact model number I designed this script for, then people can change it around however they need to?